### DISCUSSION TITLE
Manage Learn -  Observation and Survey Template Migration Plan to Sunbird Knowlg data store using KP APIs

### DISCUSSION THREAD

**Discussion by aks30** [Sunbird-Ed/Community#36] [General Discussion]
**Purpose:**

Come up with a plan to migrate existing Observation and Survey templates from Manage Learn Database to Sunbird Knowlg data store using KP APIs

**Context:**

So far Observations(with and w/o rubric) and surveys are created using Manage Learn APIs exposed for the implementation team(the partners raise tickets for the same and give the necessary data in a CSV). These questions are then consumed using Manage learn player both on the app and portal.

We have during the past few months enabled creation of observation and surveys using the creation portal(KP) thereby moving towards having all the questions in one common storage on the platform as well as removing the dependency on the implementation team. We are also working to migrate all the old and current observation and survey questions to KP. Going forward all new question sets will be created in KP only.

This document talks about how the plan to migrate existing observation and survey templates to Sunbird Knowlg datastore using KP APIs.

**Solution:**

Provide a migration script to perform the migration activity.

**Implementation Design:**

- Map the current observation and survey metadata with the sourcing program, questionset, question metadata
- Call the KP API to create program, questionset, question
- Update mongo database object with the migratedId. This will also handle failure scenarios.

**API Endpoints to use for Token Generation:**

- Dev: - https://dev.sunbirded.org/auth/realms/sunbird/protocol/openid-connect/token
- Dock: - https://dock.sunbirded.org/auth/realms/sunbird/protocol/openid-connect/token

**API Endpoints to use for Program Creation:**

- program_create: https://dock.sunbirded.org/api/program/v1/create
- program_update: https://dock.sunbirded.org/api/program/v1/update
- program_publish: https://dock.sunbirded.org/api/program/v1/publish
- add_program_nomination: https://dock.sunbirded.org/api/program/v1/nomination/add
- update_program_nomination: https://dock.sunbirded.org/api/program/v1/nomination/update

**API Endpoints to use for Question and Question Set Creation:**

- create_questionset: https://dock.sunbirded.org/api/questionset/v1/create
- update_hierarchy:https://dock.sunbirded.org/api/questionset/v1/hierarchy/update
- publish_questionset:https://dock.sunbirded.org/api/questionset/v1/publish/do_1135564342604267521239
- read_questionset : https://dock.sunbirded.org/api/questionset/v1/hierarchy/do_11357501077448294412651?mode=edit
- Question
- create_question: https://dock.sunbirded.org/api/question/v1/create
- publish_question:  https://dock.sunbirded.org/api/question/v1/publish

**API Endpoints to use for User Search, to map the templates to original creators of these templates in Manage Learn:**

- search_user: https://dev.sunbirded.org/api/user/v3/search

**Detail Execution Plan - For all observations and surveys used in Manage Learn Programs:**

Will update soon

Consumption Side API changes - [here](https://github.com/orgs/Sunbird-inQuiry/discussions/18)

Cc - @vijiurs, @kiranharidas187 @aishwaryashikshalokam, @priyanka-TL, @VISHNUDAS-tunerlabs, @Shakthieshwari, @Prateek-slokam, @MeghnaSL, @joffinjoy, @githubpiyush

**Comment by pramodkvarma** [comment]
This is much needed. Btw I was under the impression that this was done a while ago!

**Reply by aks30** [reply]
@pramodkvarma We have been working towards building/enhancing the capability on the InQuiry BB for the past few months and are now aiming to complete this exercise as part of 5.0
